light/client.go

```diff
@@ -505,151 +445,150 @@ func (c *Client) VerifyHeaderAtHeight(height int64, now time.Time) (*types.Signe
 // https://github.com/tendermint/spec/blob/master/spec/consensus/light-client.md
 //
 // If the header, which is older than the currently trusted header, is
 // requested and the light client does not have it, VerifyHeader will perform:
-//		a) bisection verification if nearest trusted header is found & not expired
+//		a) verifySkipping verification if nearest trusted header is found & not expired
 //		b) backwards verification in all other cases
 //
 // It returns ErrOldHeaderExpired if the latest trusted header expired.
 //
 // If the primary provides an invalid header (ErrInvalidHeader), it is rejected
 // and replaced by another provider until all are exhausted.
 //
-// If, at any moment, SignedHeader or ValidatorSet are not found by the primary
-// provider, provider.ErrSignedHeaderNotFound /
+// If, at any moment, lightBlock or ValidatorSet are not found by the primary
+// provider, provider.ErrlightBlockNotFound /
 // provider.ErrValidatorSetNotFound error is returned.
-func (c *Client) VerifyHeader(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+func (c *Client) VerifyHeader(newHeader *types.Header, now time.Time) error {
+	if newHeader == nil {
+		return errors.New("nil header")
+	}
 	if newHeader.Height <= 0 {
 		return errors.New("negative or zero height")
 	}
 
 	// Check if newHeader already verified.
-	h, err := c.TrustedHeader(newHeader.Height)
+	l, err := c.TrustedLightBlock(newHeader.Height)
 	if err == nil {
 		// Make sure it's the same header.
-		if !bytes.Equal(h.Hash(), newHeader.Hash()) {
-			return fmt.Errorf("existing trusted header %X does not match newHeader %X", h.Hash(), newHeader.Hash())
+		if !bytes.Equal(l.Hash(), newHeader.Hash()) {
+			return fmt.Errorf("existing trusted header %X does not match newHeader %X", l.Hash(), newHeader.Hash())
 		}
 		c.logger.Info("Header has already been verified",
 			"height", newHeader.Height, "hash", hash2str(newHeader.Hash()))
 		return nil
 	}
 
-	return c.verifyHeader(newHeader, newVals, now)
+	// Request the header and the vals.
+	l, err = c.lightBlockFromPrimary(newHeader.Height)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve light block from primary to verify against: %w", err)
+	}
+
+	if !bytes.Equal(l.Hash(), newHeader.Hash()) {
+		return fmt.Errorf("light block header %X does not match newHeader %X", l.Hash(), newHeader.Hash())
+	}
+
+	return c.verifyLightBlock(l, now)
 }
 
-func (c *Client) verifyHeader(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
-	c.logger.Info("VerifyHeader", "height", newHeader.Height, "hash", hash2str(newHeader.Hash()),
-		"vals", hash2str(newVals.Hash()))
+func (c *Client) verifyLightBlock(newLightBlock *types.LightBlock, now time.Time) error {
+	c.logger.Info("VerifyHeader", "height", newLightBlock.Height, "hash", hash2str(newLightBlock.Hash()))
+
+	var (
+		verifyFunc func(*types.LightBlock, *types.LightBlock, time.Time) error
```

Comment:
two light blocks? maybe add names to clarify... `trusted *types.LightBlock, unverified *types.LightBlock`